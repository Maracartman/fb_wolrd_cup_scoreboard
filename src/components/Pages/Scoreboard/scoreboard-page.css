.scoreboard-page {
    display: flex !important;
    width: 80%;
    background-color: transparent;
    justify-content: center;
    flex-direction: column;
    align-items: center;
}

.scoreboard-container {
   width: 100%;
   justify-content: center;
}


.summary-container{
    margin-top: 2rem;
    display: flex;
    justify-content: space-around;
    flex-direction: column;
    flex-wrap: wrap;
}

.summary-row-container {
    display: flex;
    flex-direction: row;
    width: 100%;
    font-size: x-large;
    justify-content: space-around;
    
}

.match-row-team-container {
    display: flex;
    width: 50%;
    justify-content: space-between;
    padding: 10px;
    margin-top: 10px;
    border-top: 1px solid rgb(4, 62, 21);
}

.flag-container {
    display: flex;
    align-self: center;
    width: 20px;
    height: 20px;
}

.country-name-container {
    display: flex;
    font-weight: bolder;
    font-size: x-large;
}

.score-container {
    display: flex;
    font-size: xx-large;
    font-family: Ozone;
    
}

.home-score-container {
    border-right: 1px solid rgb(4, 62, 21);   
}

.control-panel {
    display: flex !important;
    width: 100%;
    flex-direction: row;
    padding: 1rem;
    justify-content: center;
    vertical-align: middle;
    align-items: center;
}