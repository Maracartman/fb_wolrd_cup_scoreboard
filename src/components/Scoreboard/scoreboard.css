.scoreboard-container {
    align-self: center;
    flex-direction: column;
    text-align: center;
    background-color: #123103;
}

.team-scores-container {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
}

.scoreboard-title {
    color: aliceblue;
    font-size: 5vw;
    font-weight: bolder;
    font-family: monospace;
    overflow-wrap: break-word;
}
