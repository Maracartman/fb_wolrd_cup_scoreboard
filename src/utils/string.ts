
export const transformToStringScore = (score: number, size?:number): string =>{
    const places = size ?? 2
    return String(score).padStart(places, '0')
}

const StringUtil = {
    transformToStringScore
}

export default StringUtil