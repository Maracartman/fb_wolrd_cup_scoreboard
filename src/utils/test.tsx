import React, { ReactElement } from 'react'
import {RenderOptions, render} from '@testing-library/react'
import TranslationsProvider from 'src/contexts/translations'

const AllTheProviders = ({children}:{children: ReactElement}): ReactElement => (
    <TranslationsProvider>
        {children}
    </TranslationsProvider>
)

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, {wrapper: AllTheProviders, ...options})

// re-export everything
export * from '@testing-library/react'

// override render method
export {customRender as render}