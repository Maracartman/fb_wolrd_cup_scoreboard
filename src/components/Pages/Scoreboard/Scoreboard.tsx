import { useEffect, useState } from 'react'
import {  Grid, Container,Button } from '@mui/material'
import { Scoreboard } from 'src/components/Scoreboard/Scoreboard'
import { Match } from 'src/entities/Match/Match'
import teamsSample from 'src/__tests__/examples/teams.json'
import * as MatchService from 'src/services/match'
import MatchList from 'src/entities/MatchList/MatchList'
import { Team, TeamScore } from 'src/types'
import { useTranslations } from 'src/contexts/translations'
import './scoreboard-page.css'

type MatchRowProps = {
    home: TeamScore,
    away: TeamScore
}

const ScoreboardPage = () => {
    const { endMatchCopy, randomMatchCopy } = useTranslations()
    const [ongoingMatch, setOngoingMatch] = useState<Match | undefined>(undefined)
    const [matchList, setMatchList] = useState<MatchList | null>(null)

    useEffect(() => {
     MatchService.getAllMatches()
        .then(matchList => setMatchList(matchList))   
    }, [])

    const MatchRow = ({
        home,
        away,
    }: MatchRowProps) => {
        return <Grid xs={12} className='summary-row-container' item>
            <div className='match-row-team-container home-score-container'>
                {home.team?.flagImage && <div className='flag-container'>
                    <img src={home.team.flagImage} alt='Home Team Flag' className='match-row-flag' />
                </div>}
                <div className='country-name-container'>
                    {home.team?.country}
                </div>
                <div className='score-container '>
                    { home.score}
                </div>
            </div>
            <div className='match-row-team-container'>
                {away.team?.flagImage && <div className='flag-container'>
                <img src={away.team.flagImage} alt='Away Team Flag' className='match-row-flag' />
                </div>}
                <div className='country-name-container'>
                    {away.team?.country}
                </div>
                <div className='score-container'>
                    { away.score}
                </div>
            </div>
        </Grid>
    }

    /**
     * Of course, this should be replaced by something like a stream service
     * that can be subscribed in order to get the scores over the the time (match)
     */
    const showARandomMatch = () => {
        const shuffledArray = Object.values(teamsSample).sort(() => 0.5 - Math.random()); 
        const [homeTeam, awayTeam]: Team[] = shuffledArray.slice(0, 2);
        const homeScore = Math.floor(Math.random() * 10) + 1
        const awayScore = Math.floor(Math.random() * 10) + 1
        const match = new Match(homeTeam, awayTeam, {
            home: homeScore,
            away: awayScore
        })
        setOngoingMatch(match)
    }

    const finishAndAddToMatchList = () => {
        const newMatchList = matchList
        /**
         * The insert use the linked list capability to determine
         * the actual position (sorted by score and match date)
         */
        newMatchList?.insert(ongoingMatch as Match)
        setMatchList(newMatchList)
        setOngoingMatch(undefined)
    }

    return <Grid container className='scoreboard-page'>
        <Container>
        <Grid xs={12} item>
            <Scoreboard match={ongoingMatch} />
        </Grid>
        </Container>
        <Grid xs={12}>
            <Container className='control-panel'>
                <Button variant='contained'color='success' onClick={showARandomMatch}>
                    { randomMatchCopy }
                </Button>

                <Button variant='text' disabled={!ongoingMatch} onClick={finishAndAddToMatchList}>
                    { endMatchCopy }
                </Button>
            </Container>
        </Grid>
        {matchList && <Container className='summary-container'>
            {matchList.toArray().map((match: Match) => <MatchRow 
                home={match.homeTeamScore} 
                away={match.awayTeamScore}/>)}
        </Container>
        }
    </Grid>
}

export default ScoreboardPage