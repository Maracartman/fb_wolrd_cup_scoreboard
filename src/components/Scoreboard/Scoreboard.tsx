import { Match } from 'src/entities/Match/Match'
import { TeamScore } from '../TeamScore/TeamScore'
import { Box, Grid } from '@mui/material'
import { useTranslations } from 'src/contexts/translations'
import './scoreboard.css'
interface ScoreboardProps {
    match?: Match
}

export const Scoreboard = ({ match }: ScoreboardProps) => {
    const { scoreboardCopy } = useTranslations()
    const homeTeamScoreProps = match?.homeTeamScore ?? {}
    const awayTeamScoreProps = match?.awayTeamScore ?? {}
    return <Grid className='scoreboard-container' xs={12}>
        <Box className='scoreboard-title'>
            { scoreboardCopy }
        </Box>
        <Box className='team-scores-container'>
            <TeamScore {...homeTeamScoreProps} home/>
            <TeamScore {...awayTeamScoreProps}/>
        </Box>
    </Grid>
}