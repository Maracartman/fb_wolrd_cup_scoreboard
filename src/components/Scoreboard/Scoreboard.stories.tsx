import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import matchSamples from 'src/__tests__/examples/matches.json'
import { Scoreboard } from './Scoreboard';
import { Match } from 'src/entities/Match/Match';

const mapSampleToMatch = (matchSample: { home: any; away: any } ): Match => new Match(matchSample.home, matchSample.away, 
    { home: matchSample.home.score, away: matchSample.away.score})

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'Example/Scoreboard',
  component: Scoreboard,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
  },
} as ComponentMeta<typeof Scoreboard>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof Scoreboard> = (args) => <Scoreboard {...args} />

export const WithNoActiveMatch = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
WithNoActiveMatch.args = {
};

export const withAMatchPassed = Template.bind({});
withAMatchPassed.args = {
    match: mapSampleToMatch(matchSamples[1])
};
