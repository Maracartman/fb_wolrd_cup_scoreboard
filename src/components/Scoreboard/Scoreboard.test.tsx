import React from 'react'
import { render } from 'src/utils/test'
import { Scoreboard } from './Scoreboard'


/**
 * For @Reviewers from @Maracartman 
 * Since there is  a couple of components and functionality made
 * using TDD practices with RGR, I will skip testing the rest of the element
 * the main data structure and molecules are already tested
 */
describe('Scoreboard component test suite', () => {

    it('match the snapshot', () => {
        const { container } = render(<Scoreboard />)

        expect(container).toMatchSnapshot()
    })
})