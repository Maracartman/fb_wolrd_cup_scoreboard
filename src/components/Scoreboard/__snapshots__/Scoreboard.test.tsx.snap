// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scoreboard component test suite match the snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-grid-xs-12 scoreboard-container css-vj1n65-MuiGrid-root"
  >
    <div
      class="scoreboard-title MuiBox-root css-0"
    >
      Scoreboard
    </div>
    <div
      class="team-scores-container MuiBox-root css-0"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg team-score-container css-1oqqzyl-MuiContainer-root"
      >
        <div
          class="ts-data-box ts-status MuiBox-root css-0"
        >
          Home
        </div>
        <div
          class="ts-data-box country-name MuiBox-root css-0"
        >
          ---
        </div>
        <div
          class="ts-data-box MuiBox-root css-0"
        >
          <div
            class="score MuiBox-root css-0"
          >
            <div
              class="score-text"
            >
              00
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg team-score-container css-1oqqzyl-MuiContainer-root"
      >
        <div
          class="ts-data-box ts-status MuiBox-root css-0"
        >
          Away
        </div>
        <div
          class="ts-data-box country-name MuiBox-root css-0"
        >
          ---
        </div>
        <div
          class="ts-data-box MuiBox-root css-0"
        >
          <div
            class="score MuiBox-root css-0"
          >
            <div
              class="score-text"
            >
              00
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
